src/utils/FileEmitter.ts

```diff
@@ -34,7 +34,7 @@ import { isPathFragment } from './relativeId';
 import { makeUnique, renderNamePattern } from './renderNamePattern';
 
 function getSourceHash(source: string | Uint8Array): string {
-	return createHash().update(source).digest('hex');
+	return createHash().update(source).digest('base64url');
```

Comment:
Any reason for choosing this over `base64`? I could not find any documentation whether this made any difference when generating hashes, but would be interested to know.

Reply:
There are some differences. `base64url` could replace some URL charactors in the `base64` to safe character, e.g. `+` gets converted to `-` because `+` in the context of URLs means "space", `/` gets converted to `_` because `/` is a path delimiter in URLs. References: https://datatracker.ietf.org/doc/html/rfc4648#section-4, https://datatracker.ietf.org/doc/html/rfc4648#section-5.

Reply:
When I created a stackblitz link, I found that stackblitz doesn't support `base64url` encoding. It's so sad 😭. https://stackblitz.com/edit/stackblitz-starters-qswhcj?file=index.js

Reply:
That is a pity because it was introduced in Node 14.18.0. But I understand now that it is the correct encoding. I would still stick with it for now, and in the worst case (i.e. they will not add it) we switch to `digest('base64').replace(/+/g,'-').replace(/\\//g, '_')`

Reply:
Technically, StackBlitz will run the WASM build, so we only need to fix it for the node-wasm build. Note that the browser build is currently broken as well as it relies on `hash.js` and that one apparently does not understand `base64url` as well.
We could also use this opportunity to switch to a faster hashing library/algorithm, but I am not sure what the go-to approach would be.

I think technically, switching the hashing algorithm is not really a breaking change if both algorithms generate url-safe base64 using the same character sets.

So this would be my preferred approach: We do the absolute minimum now. This could be something like `digest('base64').replace(/+/g,'-').replace(/\\//g, '_')` for all environments.

Then we create an improvement ticket to switch to a faster algorithm. One approach could be to actually a really fast Rust library and expose our hashing function from Rust. What do you think?

Reply:
I had a look at `hash.js`, it only understand `hex`. I am looking for another package that support `base64` or `base64url`, If there is no such package, we have to write some hack code for `hash.js` such as changing its `digest` function.

> So this would be my preferred approach: We do the absolute minimum now. This could be something like `digest('base64').replace(/+/g,'-').replace(/\\//g, '_')` for all environments.

Yeah, It is a nice resolution for now.

> Then we create an improvement ticket to switch to a faster algorithm. One approach could be to actually a really fast Rust library and expose our hashing function from Rust. What do you think?

The `node:crypto` package is a native module that written in C++, I think it is fast enough. However, using Rust library is good to unify the crypto logical in all environments, also it can improve the performace of the rollup browser build.

Reply:
> The node:crypto package is a native module that written in C++

Yes, but its algorithms are optimized for cryptographic security, not speed. As I understand, there are faster algorithms out there when you do not care about security (as we do), see #4626

Reply:
Ah yes. So we would need to reduce the maximum hash length to 21 now to avoid a breaking change later.
Maybe it would not be too much work to do it right now and get the speed boost into the release?
Googling finds e.g. this one https://github.com/stusmall/murmur3 but there are others as well. We could go for the x64 optimised version as I would expect that more people have an x64-like system (and apparently bot versions compile on all systems). Then we expose a function via napi/wasmpack to hash a string to base 64 and also do any character conversions in Rust. Would you want to give it a shot?

Reply:
Ok, I will give it a shot.

Reply:
It's finished. Also I did some benchmarks, the result is unexpected!

<img width="1103" alt="hyperfine" src="https://github.com/rollup/rollup/assets/33021497/11e31c15-d19c-4922-93e0-b44c39a52555">

<img width="830" alt="node" src="https://github.com/rollup/rollup/assets/33021497/9c82a178-8ec7-42c5-b0ef-a467f6e9bdd2">

I found that addons maybe slower than js, some references: https://github.com/nodejs/node-addon-api/issues/790

but the wasm file is faster than addon, and it is also faster than `node:crypto`(maybe that caused by its sha256 hashing algorithm)

Reply:
Wow, that was quite some digging. I see you also discovered a much better hashing algorithm. I wonder one thing, though: Why is Node:crypto still so fast? Don't they also need to pass a JavaScript string to a native addon? Or do they have access to some dark magic that we are missing?
Question: How does performance change if you convert the string to a Buffer via `Buffer.from(string)` and pass and hash that Buffer instead? Does that improve Napi performance? Does it make a difference to use `Buffer.from(string, 'utf16le')` and hash that?

Reply:
> I wonder one thing, though: Why is Node:crypto still so fast? Don't they also need to pass a JavaScript string to a native addon? Or do they have access to some dark magic that we are missing?

I had a look at node.js, I found that the loading logic is different between `addon` and `node:crypto`, `addon` is loaded by `process.dlopen`, but `node:cypto` is loaded by `internalBinding`, is this a possible reason?

> Question: How does performance change if you convert the string to a Buffer via `Buffer.from(string)` and pass and hash that Buffer instead? Does that improve Napi performance? Does it make a difference to use `Buffer.from(string, 'utf16le')` and hash that?
<img width="1102" alt="all" src="https://github.com/rollup/rollup/assets/33021497/2ef74eb2-8da6-4245-bd64-654251702462">
<img width="1102" alt="node" src="https://github.com/rollup/rollup/assets/33021497/01d501f9-18cf-470f-9fee-2cdf5a6ff9f0">

There isn’t a significant difference between `string` and `Buffer.from(string)`. But `Buffer.from(string, 'utf16le')` is noticeably slower.

Reply:
> I had a look at node.js, I found that the loading logic is different between addon and node:crypto, addon is loaded by process.dlopen, but node:cypto is loaded by internalBinding, is this a possible reason?

Maybe, but isn't loading happening only once? Or is it calling process.dlopen for every CALL to the hashing function? In your test, how often is the function called?

Otherwise, this looks slightly disappointing. For the long-term perspective, it would be good to have the hashing implementation in Rust as then, we can easily move calls from JS to Rust. But this is not relevant now. So if we really want to be fast, we could use `xxhash-wasm` for all Node versions, keeping it as an external dependency, but use either a Rust version or maybe just `xxhash-js` for the browser build. What do you think? We use the 128bit version and limit hash length to 21 characters so we should have all breaking changes covered. And we might need to convert to base64 manually if it is not provided out-of-the-box.

Reply:
> Maybe, but isn't loading happening only once? Or is it calling process.dlopen for every CALL to the hashing function?
I think only once but the data communication maybe is different.

> how often is the function called?
1000

Reply:
> Otherwise, this looks slightly disappointing. For the long-term perspective, it would be good to have the hashing implementation in Rust as then, we can easily move calls from JS to Rust. But this is not relevant now. So if we really want to be fast, we could use `xxhash-wasm` for all Node versions, keeping it as an external dependency, but use either a Rust version or maybe just `xxhash-js` for the browser build. What do you think? We use the 128bit version and limit hash length to 21 characters so we should have all breaking changes covered. And we might need to convert to base64 manually if it is not provided out-of-the-box.

Currently, `xxhash-wasm` generates a hash with a length of 32-bit or 64-bit, `xxhash.js`(actually is ohash) generates a hash with a length of 32-bit. If we want to have the hashing implementation in Rust, I think we can use https://crates.io/crates/xxhash-rust to generate a hash with a length of 128-bit, and import its WASM files for all environments. I just tried `xxhash-rust` on my local machine.
<img width="1102" alt="1" src="https://github.com/rollup/rollup/assets/33021497/a7a5cef2-e429-4154-b1be-9bfadf5b749c">

Reply:
That would be nice, but I am not sure we want to have both a native and a WASM artifact for the non-WASM Node build. At least I assume that would very much complicate the build process.

Reply:
Perhaps now we could use the native artifact of `xxhash-rust` for the non-WASM Node build. I think it is not a big performance overhead 😊.

Reply:
Yes, let's go with this. We cannot sell it as a performance improvement yet, but it is simple for us and we have the breaking changes covered 👍.